subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java

```diff
@@ -1314,17 +1314,12 @@ public Object passThrough(Object object) {
 
     @Override
     public <T> NamedDomainObjectContainer<T> container(Class<T> type) {
-        InstantiatorFactory instantiatorFactory = getServices().get(InstantiatorFactory.class);
-        // TODO - this should also be using the decorating instantiator but cannot for backwards compatibility
-        ReflectiveNamedDomainObjectFactory<T> objectFactory = new ReflectiveNamedDomainObjectFactory<T>(type, instantiatorFactory.injectLenient(getServices()));
-        Instantiator instantiator = instantiatorFactory.decorateLenient();
-        return instantiator.newInstance(FactoryNamedDomainObjectContainer.class, type, instantiator, new DynamicPropertyNamer(), objectFactory, MutationGuards.of(getProjectConfigurator()), services.get(CollectionCallbackActionDecorator.class));
+        return getServices().get(DomainObjectCollectionFactory.class).newNamedDomainObjectContainer(type);
     }
 
     @Override
     public <T> NamedDomainObjectContainer<T> container(Class<T> type, NamedDomainObjectFactory<T> factory) {
-        Instantiator instantiator = getServices().get(InstantiatorFactory.class).decorateLenient();
-        return instantiator.newInstance(FactoryNamedDomainObjectContainer.class, type, instantiator, new DynamicPropertyNamer(), factory, MutationGuards.of(getProjectConfigurator()), services.get(CollectionCallbackActionDecorator.class));
+        return getServices().get(DomainObjectCollectionFactory.class).newNamedDomainObjectContainer(type, factory);
```

Comment:
It would be good to also move the `container(type, closure)` method to the new factory too, so that the knowledge for which implementation to use is in one place